src/File.php

```diff
@@ -45,6 +58,9 @@ public function end() : self
         if (isset($this->lastCollection)) {
             $this->lastCollection->end();
         }
+        if (isset($this->lastValue)) {
+            $this->lastValue->end();
+        }
```

Comment:
If you add a collection and then a value, and ends it, it will cause two commas wouldn't it?

I don't think having two properties `lastCollection` and `lastValue` is a good idea. What if we have a single property named `lastValue` that accepts both `Collection` and `Value`.  Then, there should be an interface that defines the `end` method. And, both `Collection` and `Value` should implement it.